File: sdk/python/hypermesh/api/trustchain.py
"""TrustChain API: certificates, issuance, validation, revocation, DNS zones."""

from __future__ import annotations

from typing import Any

from ..types import (
    TrustChainCertificate,
    TrustChainCertificateList,
    TrustChainDnsZoneList,
    TrustChainRevokeResult,
    TrustChainValidationResult,
)

_PREFIX = "/api/v1/trustchain"


class TrustChainApi:
    """Wraps /api/v1/trustchain/* endpoints."""

    def __init__(self, transport: Any) -> None:
        self._t = transport

    def certificates(self) -> TrustChainCertificateList:
        data = self._t.get(f"{_PREFIX}/certificates")
        return _parse_certificate_list(data)

    def issue(self, subject: str, scope: str) -> TrustChainCertificate:
        data = self._t.post(
            f"{_PREFIX}/issue", {"subject": subject, "scope": scope}
        )
        return _parse_certificate(data)

    def validate(self, cert_pem: str) -> TrustChainValidationResult:
        data = self._t.post(f"{_PREFIX}/validate", {"cert_pem": cert_pem})
        return TrustChainValidationResult(
            valid=data.get("valid", False),
            errors=data.get("errors", []),
            chain_valid=data.get("chain_valid", False),
        )

    def revoke(self, cert_id: str) -> TrustChainRevokeResult:
        data = self._t.post(f"{_PREFIX}/revoke", {"cert_id": cert_id})
        return TrustChainRevokeResult(
            revoked=data.get("revoked", False),
            cert_id=data.get("cert_id", cert_id),
        )

    def dns_zones(self) -> TrustChainDnsZoneList:
        data = self._t.get(f"{_PREFIX}/dns/zones")
        return _parse_dns_zone_list(data)


class AsyncTrustChainApi:
    """Async variant of TrustChainApi."""

    def __init__(self, transport: Any) -> None:
        self._t = transport

    async def certificates(self) -> TrustChainCertificateList:
        data = await self._t.get(f"{_PREFIX}/certificates")
        return _parse_certificate_list(data)

    async def issue(self, subject: str, scope: str) -> TrustChainCertificate:
        data = await self._t.post(
            f"{_PREFIX}/issue", {"subject": subject, "scope": scope}
        )
        return _parse_certificate(data)

    async def validate(self, cert_pem: str) -> TrustChainValidationResult:
        data = await self._t.post(
            f"{_PREFIX}/validate", {"cert_pem": cert_pem}
        )
        return TrustChainValidationResult(
            valid=data.get("valid", False),
            errors=data.get("errors", []),
            chain_valid=data.get("chain_valid", False),
        )

    async def revoke(self, cert_id: str) -> TrustChainRevokeResult:
        data = await self._t.post(f"{_PREFIX}/revoke", {"cert_id": cert_id})
        return TrustChainRevokeResult(
            revoked=data.get("revoked", False),
            cert_id=data.get("cert_id", cert_id),
        )

    async def dns_zones(self) -> TrustChainDnsZoneList:
        data = await self._t.get(f"{_PREFIX}/dns/zones")
        return _parse_dns_zone_list(data)


def _parse_certificate(data: dict[str, Any]) -> TrustChainCertificate:
    return TrustChainCertificate(
        id=data.get("id", ""),
        subject=data.get("subject", ""),
        scope=data.get("scope", ""),
        valid_from=data.get("valid_from", ""),
        valid_to=data.get("valid_to", ""),
        pem=data.get("pem", ""),
    )


def _parse_certificate_list(
    data: dict[str, Any],
) -> TrustChainCertificateList:
    certs = [_parse_certificate(c) for c in data.get("certificates", [])]
    return TrustChainCertificateList(
        count=data.get("count", len(certs)),
        certificates=certs,
    )


def _parse_dns_zone_list(data: dict[str, Any]) -> TrustChainDnsZoneList:
    zones = data.get("zones", [])
    return TrustChainDnsZoneList(
        count=data.get("count", len(zones)),
        zones=zones,
    )

File: sdk/go/trustchain.go
package hypermesh

import "context"

// TrustChainApi provides access to TrustChain endpoints.
type TrustChainApi struct {
	http *httpClient
}

// Certificates returns all certificates.
func (a *TrustChainApi) Certificates(ctx context.Context) (*TrustChainCertificateList, error) {
	var result TrustChainCertificateList
	if err := a.http.get(ctx, "/api/v1/trustchain/certificates", &result); err != nil {
		return nil, err
	}
	return &result, nil
}

// Issue issues a new certificate for the given subject and scope.
func (a *TrustChainApi) Issue(ctx context.Context, subject, scope string) (*TrustChainCertificate, error) {
	body := TrustChainIssueRequest{Subject: subject, Scope: scope}
	var result TrustChainCertificate
	if err := a.http.post(ctx, "/api/v1/trustchain/issue", body, &result); err != nil {
		return nil, err
	}
	return &result, nil
}

// Validate validates the given certificate PEM.
func (a *TrustChainApi) Validate(ctx context.Context, certPem string) (*TrustChainValidationResult, error) {
	body := TrustChainValidateRequest{CertPem: certPem}
	var result TrustChainValidationResult
	if err := a.http.post(ctx, "/api/v1/trustchain/validate", body, &result); err != nil {
		return nil, err
	}
	return &result, nil
}

// Revoke revokes the certificate with the given ID.
func (a *TrustChainApi) Revoke(ctx context.Context, certID string) (*TrustChainRevokeResult, error) {
	body := TrustChainRevokeRequest{CertID: certID}
	var result TrustChainRevokeResult
	if err := a.http.post(ctx, "/api/v1/trustchain/revoke", body, &result); err != nil {
		return nil, err
	}
	return &result, nil
}

// DnsZones returns all DNS zones.
func (a *TrustChainApi) DnsZones(ctx context.Context) (*TrustChainDnsZoneList, error) {
	var result TrustChainDnsZoneList
	if err := a.http.get(ctx, "/api/v1/trustchain/dns/zones", &result); err != nil {
		return nil, err
	}
	return &result, nil
}
